src/dht_proxy_client.cpp

```diff
@@ -0,0 +1,587 @@
+    }
+}
+
+bool
+DhtProxyClient::isRunning(sa_family_t af) const
+{
+    switch (af)
+    {
+    case AF_INET:
+        return statusIpv4_ == NodeStatus::Connected;
+    case AF_INET6:
+        return statusIpv6_ == NodeStatus::Connected;
+    default:
+        return false;
+    }
+}
+
+time_point
+DhtProxyClient::periodic(const uint8_t*, size_t, const SockAddr&)
+{
+    // Exec all currently stored callbacks
+    scheduler.syncTime();
+    if (!callbacks_.empty()) {
+        lockCallbacks.lock();
+        for (auto& callback : callbacks_)
+            callback();
+        callbacks_.clear();
+        lockCallbacks.unlock();
+    }
+    // Remove finished operations
+    lockOperations_.lock();
+    auto operation = operations_.begin();
+    while (operation != operations_.end()) {
+        if (*(operation->finished)) {
+            if (operation->thread.joinable()) {
+                // Close connection to stop operation?
+                restbed::Http::close(operation->req);
+                operation->thread.join();
+            }
+            operation = operations_.erase(operation);
+        } else {
+            ++operation;
+        }
+    }
+    lockOperations_.unlock();
+    return scheduler.run();
+}
+
+void
+DhtProxyClient::get(const InfoHash& key, GetCallback cb, DoneCallback donecb,
+                    Value::Filter&& filter, Where&& where)
+{
+    restbed::Uri uri(HTTP_PROTO + serverHost_ + "/" + key.toString());
+    auto req = std::make_shared<restbed::Request>(uri);
+    Query query {{}, where};
+    auto filterChain = filter.chain(query.where.getFilter());
+
+    auto finished = std::make_shared<bool>(false);
+    Operation o;
+    o.req = req;
+    o.finished = finished;
+    o.thread = std::move(std::thread([=](){
+        // Try to contact the proxy and set the status to connected when done.
+        // will change the connectivity status
+        auto ok = std::make_shared<bool>(true);
+        restbed::Http::async(req,
+            [=](const std::shared_ptr<restbed::Request>& req,
+                const std::shared_ptr<restbed::Response>& reply) {
+            auto code = reply->get_status_code();
+
+            if (code == 200) {
+                try {
+                    while (restbed::Http::is_open(req)) {
+                        restbed::Http::fetch("\n", reply);
+                        std::string body;
+                        reply->get_body(body);
+                        reply->set_body(""); // Reset the body for the next fetch
+
+                        Json::Value json;
+                        Json::Reader reader;
+                        if (reader.parse(body, json)) {
+                            auto value = std::make_shared<Value>(json);
+                            if ((not filterChain or filterChain(*value)) && cb) {
+                                lockCallbacks.lock();
+                                callbacks_.emplace_back([=](){
+                                    cb({value});
```

Comment:
if cb returns false it should stop the operation and call donecb

Reply:
Ok.